rgw/admin/user.go

```diff
@@ -0,0 +1,148 @@
+		User      string `json:"user"`
+		AccessKey string `json:"access_key" url:"access-key"`
+		SecretKey string `json:"secret_key" url:"secret-key"`
+	} `json:"keys"`
+	SwiftKeys []interface{} `json:"swift_keys"`
+	Caps      []struct {
+		Type string `json:"type"`
+		Perm string `json:"perm"`
+	} `json:"caps"`
+	OpMask              string        `json:"op_mask"`
+	DefaultPlacement    string        `json:"default_placement"`
+	DefaultStorageClass string        `json:"default_storage_class"`
+	PlacementTags       []interface{} `json:"placement_tags"`
+	BucketQuota         struct {
+		Enabled    bool `json:"enabled"`
+		CheckOnRaw bool `json:"check_on_raw"`
+		MaxSize    int  `json:"max_size"`
+		MaxSizeKb  int  `json:"max_size_kb"`
+		MaxObjects int  `json:"max_objects"`
+	} `json:"bucket_quota"`
+	UserQuota struct {
+		Enabled    bool `json:"enabled"`
+		CheckOnRaw bool `json:"check_on_raw"`
+		MaxSize    int  `json:"max_size"`
+		MaxSizeKb  int  `json:"max_size_kb"`
+		MaxObjects int  `json:"max_objects"`
+	} `json:"user_quota"`
+	TempURLKeys []interface{} `json:"temp_url_keys"`
+	Type        string        `json:"type"`
+	MfaIds      []interface{} `json:"mfa_ids"`
+	KeyType     string        `url:"key-type"`
+	Tenant      string        `url:"tenant"`
+	GenerateKey bool          `url:"generate-key"`
+	PurgeData   int           `url:"purge-data"`
+}
+
+// GetUser retrieves a given object store user
+func (api *API) GetUser(ctx context.Context, user User) (*User, error) {
+	if user.UserID == "" {
+		return nil, errors.New("missing user UID")
+	}
+
+	body, err := api.callAPI(ctx, get, "/user", getValues(user))
+	if err != nil {
+		return nil, err
+	}
+
+	u := &User{}
+	err = json.Unmarshal(body, &u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal radosgw user http response. %s. %w", string(body), err)
+	}
+
+	return u, nil
+}
+
+// GetUsers lists all object store users
+func (api *API) GetUsers(ctx context.Context) (*[]string, error) {
+	body, err := api.callAPI(ctx, get, "/metadata/user", nil)
+	if err != nil {
+		return nil, err
+	}
+	var users *[]string
+	err = json.Unmarshal(body, &users)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal radosgw user http response. %s. %w", string(body), err)
+	}
+
+	return users, nil
+}
+
+// CreateUser creates a user in the object store
+func (api *API) CreateUser(ctx context.Context, user User) (*User, error) {
+	if user.UserID == "" {
+		return nil, errors.New("missing user UID")
+	}
+	if user.DisplayName == "" {
+		return nil, errors.New("missing user display name")
+	}
+
+	// Send request
+	body, err := api.callAPI(ctx, put, "/user", getValues(user))
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal response into Go type
+	u := &User{}
+	err = json.Unmarshal(body, &u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal radosgw user http response. %s. %w", string(body), err)
+	}
+
+	return u, nil
+}
+
+// RemoveUser remove an user from the object store
```

Comment:
for `RemoveUser()` and `ModifyUser()` IMO it is better to add check whether they are performing this operation on the `User` which the credentials passed in `New()` request. I know normally users won't do it and if they do it is a clear mistake from their side.

Reply:
I don't understand, can you clarify?

Reply:
For the admin API to work we need to create s3 user's credentials, the 'ModifyUser()' or 'DeleteUser()' should touch that user IMO.